ports/rp2/machine_i2c.c

```diff
@@ -0,0 +1,160 @@
+};
+
+STATIC void machine_i2c_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_i2c_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "I2C(%u, freq=%u, scl=%u, sda=%u)",
+        self->i2c_id, self->freq, self->scl, self->sda);
+}
+
+mp_obj_t machine_i2c_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_id, ARG_freq, ARG_scl, ARG_sda };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_freq, MP_ARG_INT, {.u_int = DEFAULT_I2C_FREQ} },
+        { MP_QSTR_scl, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
+        { MP_QSTR_sda, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
+    };
+
+    // Parse args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Get I2C bus.
+    int i2c_id = mp_obj_get_int(args[ARG_id].u_obj);
+    if (i2c_id < 0 || i2c_id >= MP_ARRAY_SIZE(machine_i2c_obj)) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("I2C(%d) doesn't exist"), i2c_id);
+    }
+
+    // Get static peripheral object.
+    machine_i2c_obj_t *self = (machine_i2c_obj_t *)&machine_i2c_obj[i2c_id];
+
+    // Set SCL/SDA pins if configured.
+    if (args[ARG_scl].u_obj != mp_const_none) {
+        int scl = mp_hal_get_pin_obj(args[ARG_scl].u_obj);
+        if (!IS_VALID_SCL(self->i2c_id, scl)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("bad SCL pin"));
+        }
+        self->scl = scl;
+    }
+    if (args[ARG_sda].u_obj != mp_const_none) {
+        int sda = mp_hal_get_pin_obj(args[ARG_sda].u_obj);
+        if (!IS_VALID_SDA(self->i2c_id, sda)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("bad SDA pin"));
+        }
+        self->sda = sda;
+    }
+
+    // Initialise the I2C peripheral if any arguments given, or it was not initialised previously.
+    if (n_args > 1 || n_kw > 0 || self->freq == 0) {
+        self->freq = args[ARG_freq].u_int;
+        i2c_init(self->i2c_inst, self->freq);
+        self->freq = i2c_set_baudrate(self->i2c_inst, self->freq);
+        gpio_set_function(self->scl, GPIO_FUNC_I2C);
+        gpio_set_function(self->sda, GPIO_FUNC_I2C);
+        gpio_set_pulls(self->scl, true, 0);
+        gpio_set_pulls(self->sda, true, 0);
+    }
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC int machine_i2c_transfer_single(mp_obj_base_t *self_in, uint16_t addr, size_t len, uint8_t *buf, unsigned int flags) {
+    machine_i2c_obj_t *self = (machine_i2c_obj_t *)self_in;
+    int ret;
+    bool nostop = !(flags & MP_MACHINE_I2C_FLAG_STOP);
+    if (flags & MP_MACHINE_I2C_FLAG_READ) {
+        ret = i2c_read_blocking(self->i2c_inst, addr, buf, len, nostop);
+    } else {
+        if (len <= 2) {
+            // Workaround issue with hardware I2C not accepting short writes.
```

Comment:
Is there an rp2040 errata or pico-sdk bug for this?

Reply:
Not sure... I got this from inspecting the SDK's behaviour.  I don't think it's a bug, just a limitation of the hardware.